rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,755 @@
+  // successful, the MFA device used for authentication is returned. If the verification fails, an error is returned.
+  rpc VerifyValidatedMFAChallenge(VerifyValidatedMFAChallengeRequest) returns (VerifyValidatedMFAChallengeResponse);
+}
+
+// SessionIdentifyingPayload contains a value that uniquely identifies a user's session.
+// It must be computed by the client from session state and is used to bind MFA challenges to specific sessions.
+message SessionIdentifyingPayload {
+  oneof payload {
+    teleport.ssh.v1.SessionPayload ssh_session = 1;
+  }
+}
+
+// CreateChallengeRequest is the request message for CreateChallenge.
+message CreateChallengeRequest {
+  // payload is a value that uniquely identifies the user's session. When VerifyValidatedMFAChallenge is called, the server
+  // will verify it matches the payload supplied to CreateChallengeRequest.
+  SessionIdentifyingPayload payload = 1;
+  // sso_client_redirect_url should be supplied if the client supports SSO MFA checks. If unset, the server will only
+  // return non-SSO challenges.
+  string sso_client_redirect_url = 2;
+  // proxy_address is the proxy address that the user is using to connect to the Proxy. When using SSO MFA, this address
+  // is required to determine which URL to redirect the user to when there are multiple options.
+  string proxy_address = 3;
+}
+
+// CreateChallengeResponse is the response message for CreateChallenge.
+message CreateChallengeResponse {
+  // name is the unique resource name for the issued challenge generated by the server. This value should be treated as
+  // an opaque identifier and used as-is in subsequent API calls.
+  string name = 1;
+  // mfa_challenge contains the MFA challenge that the user must respond to.
+  AuthenticateChallenge mfa_challenge = 2;
+}
+
+// ValidateChallengeRequest is the request message for ValidateChallenge.
+message ValidateChallengeRequest {
+  // name is the resource name for the issued challenge.
+  // This must match the 'name' returned in CreateChallengeResponse to tie the validation to the correct challenge.
+  string name = 1;
+  // mfa_response contains the MFA challenge response provided by the user.
+  AuthenticateResponse mfa_response = 2;
+}
+
+// ValidateChallengeResponse is the response message for ValidateChallenge.
+message ValidateChallengeResponse {}
+
+// ReplicateValidatedMFAChallengeRequest is the request message for ReplicateValidatedMFAChallenge.
+message ReplicateValidatedMFAChallengeRequest {
+  // name is the resource name for the issued challenge.
+  // This must match the 'name' returned in CreateChallengeResponse to tie the upsert to the correct challenge.
+  string name = 1;
+  // payload is a value that uniquely identifies the user's session. This should match the payload in
+  // CreateChallengeRequest.
+  SessionIdentifyingPayload payload = 2;
+  // device contains information about the user's MFA device used to authenticate.
+  types.MFADevice device = 3;
+}
```

Comment:
Missing source and target clusters, even if we enforce that the former matches the identity of the caller and the destination is the name of the local cluster it might save us from a bug if we get the replication logic wrong